vault/token_store.go

```diff
@@ -1078,72 +1077,56 @@ func (ts *TokenStore) Revoke(ctx context.Context, id string) error {
 // revokeSalted is used to invalidate a given salted token,
 // any child tokens will be orphaned.
 func (ts *TokenStore) revokeSalted(ctx context.Context, saltedID string) (ret error) {
-	// Protect the entry lookup/writing with locks. The rub here is that we
-	// don't know the ID until we look it up once, so first we look it up, then
-	// do a locked lookup.
-	entry, err := ts.lookupSalted(ctx, saltedID, true)
-	if err != nil {
-		return err
-	}
-	if entry == nil {
+	// Check and set the token deletion state. We only proceed with the deletion
+	// if we don't have a pending deletion (empty), or if the deletion previously
+	// failed (state is false)
+	state, loaded := ts.tokensPendingDeletion.LoadOrStore(saltedID, true)
+
+	// If the entry was loaded and it's state is true, we shortcircuit
+	if loaded && state == true {
 		return nil
 	}
 
-	lock := locksutil.LockForKey(ts.tokenLocks, entry.ID)
-	lock.Lock()
-
-	// Lookup the token first
-	entry, err = ts.lookupSalted(ctx, saltedID, true)
+	// The map check above should protect use from any concurrent revocations, so
+	// doing a bare lookup here should be fine.
+	entry, err := ts.lookupSalted(ctx, saltedID, true)
 	if err != nil {
-		lock.Unlock()
 		return err
 	}
-
 	if entry == nil {
-		lock.Unlock()
-		return nil
-	}
-
-	// On failure we write -3, so if we hit -2 here we're already running a
-	// revocation operation. This can happen due to e.g. recursion into this
-	// function via the expiration manager's RevokeByToken.
-	if entry.NumUses == tokenRevocationInProgress {
-		lock.Unlock()
 		return nil
 	}
 
-	// This acts as a WAL. lookupSalted will no longer return this entry,
-	// so the token cannot be used, but this way we can keep the entry
-	// around until after the rest of this function is attempted, and a
-	// tidy function can key off of this value to try again.
-	entry.NumUses = tokenRevocationInProgress
-	err = ts.storeCommon(ctx, entry, false)
-	lock.Unlock()
-	if err != nil {
-		return err
+	if entry.NumUses != tokenRevocationPending {
```

Comment:
Do we need this `if` check?

Reply:
It'd make sense to keep this `if` for the case that the token is already in this state so that we don't have to re-store it with the same value.